src/Acuminator/Acuminator.Analyzers/StaticAnalysis/ViewDeclarationOrder/ViewDeclarationOrderAnalyzer.cs

```diff
@@ -1,77 +1,124 @@
+	/// <summary>
+	/// An analyzer for the order of view declaration in graph/graph extension.
+	/// This diagnostic works only for simple class hierarchy where the depth of the inheritance is equal to 2: object -> DAC -> DerivedDAC.
+	/// </summary>
+	public class ViewDeclarationOrderAnalyzer : IPXGraphAnalyzer
+	{
+		private const string InitCacheMappingMethodName = "InitCacheMapping";
 
-		internal override void AnalyzeCompilation(CompilationStartAnalysisContext compilationStartContext, PXContext pxContext)
-		{
-            compilationStartContext.RegisterSymbolAction(c => AnalyzeGraphViews(c, pxContext), SymbolKind.NamedType);
-		}
+		public ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(Descriptors.PX1004_ViewDeclarationOrder, Descriptors.PX1006_ViewDeclarationOrder);
 
-		private static void AnalyzeGraphViews(SymbolAnalysisContext context, PXContext pxContext)
+		public void Analyze(SymbolAnalysisContext symbolContext, PXContext pxContext, CodeAnalysisSettings settings,
+							PXGraphSemanticModel graphSemanticModel)
 		{
-			INamedTypeSymbol graph = (INamedTypeSymbol)context.Symbol;
-
-			if (graph == null || context.CancellationToken.IsCancellationRequested ||
-			   !graph.InheritsFromOrEquals(pxContext.PXGraph.Type) && !graph.InheritsFromOrEquals(pxContext.PXGraphExtensionType))
-			{
-				return;
-			}
-        
-            var graphViews = graph.GetMembers()
-								  .OfType<IFieldSymbol>()
-								  .Select(field => field.Type as INamedTypeSymbol)
-								  .Where(fieldType => fieldType != null &&
-													   fieldType.InheritsFrom(pxContext.PXSelectBase.Type) && 
-													   fieldType.IsGenericType &&
-													   fieldType.TypeArguments.Length > 0).
-								   ToImmutableList();
-
-			Location graphLocation = graph.Locations.FirstOrDefault();
-
-			if (graphLocation == null || context.CancellationToken.IsCancellationRequested)
+			if (graphSemanticModel.ViewsByNames.Count == 0 || IsNewMethodUsedToInitCaches(pxContext))
 				return;
 
-			//forward pass
-			CheckGraphViews(context, graphLocation, graphViews, Descriptors.PX1004_ViewDeclarationOrder);
+			symbolContext.CancellationToken.ThrowIfCancellationRequested();
 
-			if (context.CancellationToken.IsCancellationRequested)
+			var viewsGroupedByDAC = GetViewsUsedInAnalysis(graphSemanticModel).Where(view => view.ViewDAC != null)
+																			  .ToLookup(view => view.ViewDAC);
+			if (viewsGroupedByDAC.Count == 0)
 				return;
 
-			//backward pass
-			CheckGraphViews(context, graphLocation, graphViews.Reverse(), Descriptors.PX1006_ViewDeclarationOrder);
+			foreach (IGrouping<ITypeSymbol, DataViewInfo> dacViews in viewsGroupedByDAC)
+			{
+				AnalyzeDacViewsForNumberOfCaches(graphSemanticModel, symbolContext, dacViews, viewsGroupedByDAC);
+			}
+
+			symbolContext.CancellationToken.ThrowIfCancellationRequested();
```

Comment:
What is the reason to check cancellation token at the end of the method? It will decrease performance.

Reply:
ok